docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,220 @@
+
+It is important to note that in this formula, :math:`\frac{\partial y}{\partial a}` and :math:`\frac{\partial y}{\partial b}` are simple :math:`\mathcal{R} \to \mathcal{R}` derivatives.
+To evaluate this numerically, we thus approximate :math:`\frac{\partial y}{\partial a}` and :math:`\frac{\partial y}{\partial b}` using the method described above for the real-to-real case, compute the :math:`CW` matrix and then multiply it by :math:`2`.
+
+Note that the code, as of time of writing, computes this value in a slightly convoluted way:
+
+.. code:: python
+
+    # Code from https://github.com/pytorch/pytorch/blob/58eb23378f2a376565a66ac32c93a316c45b6131/torch/autograd/gradcheck.py#L99-L105
+    # Notation changes in this code block:
+    # s here is y above
+    # x, y here are a, b above
+
+    ds_dx = compute_gradient(eps)
+    ds_dy = compute_gradient(eps * 1j)
+    # conjugate wirtinger derivative
+    conj_w_d = 0.5 * (ds_dx + ds_dy * 1j)
+    # wirtinger derivative
+    w_d = 0.5 * (ds_dx - ds_dy * 1j)
+    d[d_idx] = grad_out.conjugate() * conj_w_d + grad_out * w_d.conj()
+
+    # Since grad_out is always 1, and W and CW are complex conjugate of each other, the last line ends up computing exactly `conj_w_d + w_d.conj() = conj_w_d + conj_w_d = 2 * conj_w_d`.
+
+
+Slow complex input analytical evaluation
+""""""""""""""""""""""""""""""""""""""""
+
+For this case, since the backward mode AD is computing exactly twice the :math:`CW` derivative already, we simply use the same trick as for the real-to-real case and reconstruct the matrix row by row when there are multiple real outputs.
+
+Functions with outputs in :math:`\mathcal{C}^M`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this case, the assumption we have about the computation of the :math:`CW` does not hold (output is not real) and so we cannot compute it directly to compare it.
+To solve this, we will replace the test of the function :math:`h: \mathcal{P}^N \to \mathcal{C}^M` (where :math:`\mathcal{P}` can be either :math:`\mathcal{R}` or :math:`\mathcal{C}`), with two functions :math:`hr` and :math:`hi` such that: :math:`hr(q) = real(f(q))` and :math:`hi(q) = imag(f(q))` where :math:`q \in \mathcal{P}`.
+We then do a basic gradcheck for both :math:`hr` and :math:`hi` using either the real-to-real or complex to real case described above, depending on :math:`\mathcal{P}`.
+
+Note that in the code, as of time of writing, does not create these functions explicitly but perform the chain rule with the :math:`real` or :math:`imag` functions manually by passing the :math:`\text{grad\_out}` arguments to the different functions.
+When :math:`\text{grad\_out} = 1`, then we are considering :math:`hr`.
+When :math:`\text{grad\_out} = i`, then we are considering :math:`hi`.
+
+
+Fast backward mode gradcheck
+----------------------------
+
+While the above formulation of gradcheck is great both to ensure correctness and debuggability, it is very slow the reconstruct full Jacobian matrices.
```

Comment:
```suggestion
While the above formulation of gradcheck is great, both, to ensure correctness and debuggability, it is very slow since it reconstructs the full Jacobian matrices.
```